include/seqan3/contrib/parallel/buffer_queue.hpp

```diff
@@ -191,46 +228,88 @@ class buffer_queue
     bool is_empty() const noexcept
     {
         std::unique_lock write_lock(mutex);
-        return headPos == tailPos;
+        return pop_front_position == push_back_position;
     }
 
     bool is_full() const noexcept
     {
         std::unique_lock write_lock(mutex);
-        return size_impl() == data.max_size();
+        return !has_free_slots(pop_front_position, push_back_position);
     }
 
     size_type size() const noexcept
     {
         std::unique_lock write_lock(mutex);
-        return size_impl();
+        size_type number_of_elements = push_back_position - pop_front_position;
+        // During a wrap-around of the `push_back_position` the term `(ring_buffer_capacity - buffer.size())` is added.
+        if (to_buffer_position(pop_front_position) > to_buffer_position(push_back_position))
+            number_of_elements -= (ring_buffer_capacity - buffer.size());
+        return number_of_elements;
     }
     //!\}
 private:
 
-    // Needed in two functions that both acquire exclusive rights.
-    size_type size_impl() const noexcept
+    /*!\brief Checks if enough slots are available for the `push_back_position` to write to the ring buffer.
+     * \param[in] pop_front_position The thread local position to read from.
+     * \param[in] push_back_position The thread local position to write to the buffer.
+     *
+     * \details
+     *
+     * The seqan3::contrib::buffer_queue::cyclic_increment ensures that the `push_back_position` is at least
+     * `ring_buffer_capacity` many slots ahead of `pop_front_position` if the queue is full.
+     */
+    constexpr bool has_free_slots(size_type const pop_front_position, size_type const push_back_position) const
```

Comment:
1. The variables `pop_front_position` and `push_back_position` are member variables, so you do not need to pass them as arguments to `has_free_slots` as they are already known...
2. This function needs a write lock, I assume.

Reply:
Well, you need to be careful here. The member functions are atomics and we use specifically the thread local versions inside of the functions here. This is important since the state of the atomic might have been changed since a particular thread read a state of the atomic until it uses it. Thus, the thread local variables are not the same as the global atomic ones.